### PR TITLE
Support 'selectionMode' on tabular and cards

### LINKED ISSUE #7818: Behaviour of 'general checkbox' button in M2O interface
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

The M2O-Interface shows a checkbox button in the upper left corner. Clicking it always just selects the last element in the list which is likely not the intended behaviour. It makes no difference if an item is already selected (so it shows a minus in the checkbox) or not (so it shows an empty checkbox).
Unbenannt

### To Reproduce

Open a M2O interface and click on the checkbox button in the upper left corner.

### What version of Directus are you using?

9.0.0-rc.92

### What version of Node.js are you using?

14.17.5

### What database are you using?

MariaDB and MySQL

### What browser are you using?

Firefox

### What operating system are you using?

Linux

### How are you deploying Directus?

local

### LINKED ISSUE #7935: Activity Listing Selection
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

Items on the activity page can not be selected, but they have checkboxes... that don't do anything. We should see if it makes sense to allow selecting items (eg: admins removing legacy activity), or if we should remove the checkboxes.

Let's discuss! 

### To Reproduce

Go to the activity page, click on an item checkbox

### What version of Directus are you using?

RC-92

### What version of Node.js are you using?

Node: 16.6.1

### What database are you using?

Postgres 13

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying Directus?

DO

### LINKED ISSUE #8019: Drawer-collection select: Only every third click lucky winner  😅
### Preflight Checklist

- [X] I have completed all [Troubleshooting Steps](https://docs.directus.io/getting-started/support/#troubleshooting-steps).
- [X] I'm on [the latest version of Directus](https://github.com/directus/directus/releases).
- [X] There's [no other issue](https://github.com/directus/directus/issues) that already describes my problem.

### Describe the Bug

When selecting an existing item for a M2O-relation in the `<drawer-collection>`-select, (in my test case from a `select-dropdown-m2o`) you can normally select the item by:

**A**, either clicking directly in the little checkbox on the left 

Bildschirmfoto 2021-09-14 um 09 44 44

or **B.**, by clicking **anywhere else in the row**, e.g. on the author name, in the image above. This is super useful, as clicking the little checkbox directly is a bit tedious.

Procedure **A**, using the little box still, works on first try, but **B** only ever works on the **third click**.

But that's not all to it:

1. This usually happens only once while selecting items. Once you did your 3 clicks, the rows become kind of "activated" and start working on first click.
2. This only happens with unselected rows. If you e.g. have an existing item selected, and want to deselect it or associate another item, the "deselection"-click works on first try!
3. Plus: When you did a deselection click, the rows again become kind of "activated" again and every click works on first try.
4. Maybe unrelated: If you come from a `select-dropdown-m2o` interface and have an existing item selected, and you select another item, the previously selected item does not become unselected. In other words: Although you cannot select more than one item, the drawer-selection interface acts like you could select multiple. If this is unrelated to the issue described here, I will open a separate one.

### To Reproduce

Just try to select in a drawer-collection or use this reproduction repo including SQLite data:
https://github.com/bluepick-app/directus-drawer-select-three-clicks-issue

I could reproduce this both, on Chrome and Safari to make sure it's not a Chrome-extensions thing.

### What version of Directus are you using?

9.0.0-rc.92

### What version of Node.js are you using?

>=14

### What database are you using?

Postgres 13

### What browser are you using?

Chrome, Safari

### What operating system are you using?

macOS

### How are you deploying Directus?

running locally

### PR DESCRIPTION
**Pull request by joselcvarela**
If 'selectionMode'
== 'no-select' : hides checkboxes
== 'select-one': hides select all but show checkboxes (as radio buttons)
== 'select-multiple': show select all and checkboxes

Fixes #7935 
Fixes #7818 
Fixes #7761 (not using v-radio, but emulating with icons)

----
Edit:
Fixes #8019

### CODE REVIEW

**Comment by joselcvarela** [comment]
Before:

https://user-images.githubusercontent.com/14039341/133156290-38f91fea-e7cc-4fda-b8fe-ab58cd374b5e.mp4

After:

https://user-images.githubusercontent.com/14039341/133156299-1db8288b-6ce3-485c-a71f-1f08c4a68c5a.mp4

**Comment by rijkvanzanten** [comment]
@joselcvarela we should also figure out how this would work for the Calendar and Map layouts 🤔

**Comment by joselcvarela** [comment]
I didn't find the select all on those layouts. I am still discovering the app, so it's completely normal. I will setup those fields and try to replicate the behavior. But my concern is how the 'selectionMode' is declared. It is ok like that or should I change?

**Comment by benhaynes** [comment]
Interesting point. I don't think we _have_ a selection mode in Map or Calendar... but we should (need to?). I can work on a design for this.

**Comment by benhaynes** [comment]
By the way, this is awesome work, @joselcvarela!! ❤️

**Comment by rijkvanzanten** [comment]
> But my concern is how the 'selectionMode' is declared. It is ok like that or should I change?

Yeah looks good to me! 

The only thing I would change is that I would call them "one, multiple, none" instead of "select-*" as we're already in the selectMode context 🙂

**Comment by joselcvarela** [comment]
Updated 😉

File: app/src/components/v-table/table-header/table-header.vue [hunk 1]
```diff
@@ -65,9 +67,9 @@ export default defineComponent({
 			type: Object as PropType<Sort>,
 			required: true,
 		},
-		showSelect: {
-			type: Boolean,
-			default: false,
+		selectionMode: {
+			type: String as PropType<LayoutProps['selectionMode']>,
```
Comment (nickrum) [inline app/src/components/v-table/table-header/table-header.vue:71]: Maybe we should introduce a separate `SelectionMode` type to separate the table component from layout related types.

This also applies to a few cases below.

File: app/src/components/v-table/table-header/table-header.vue [hunk 2]
```diff
@@ -114,6 +116,8 @@ export default defineComponent({
 		useEventListener(window, 'pointermove', throttle(onMouseMove, 40));
 		useEventListener(window, 'pointerup', onMouseUp);
 
+		const showSelect = ['one', 'multiple'].includes(props.selectionMode);
```
Comment (nickrum) [inline app/src/components/v-table/table-header/table-header.vue:119]: This should probably be a computed property in case the prop changes.

This also applies to a few cases below.

File: app/src/modules/settings/routes/presets/collection/collection.vue [hunk 3]
```diff
@@ -56,7 +56,7 @@
 				fixed-header
 				:items="presets"
 				:loading="loading"
-				show-select
+				:selection-mode="'multiple'"
```
Comment (nickrum) [inline app/src/modules/settings/routes/presets/collection/collection.vue:59]:
```suggestion
				selection-mode="multiple"
```

**Review by nickrum — CHANGES_REQUESTED** [review]
With this PR we have `selection`, `selectMode` and `selectionMode` layout props. This might be a bit confusing, but I can't think of a better name either.

**Comment by rijkvanzanten** [comment]
>  selection, selectMode and selectionMode 

Yeah we should differentiate between those last two somehow 🤔

**Comment by joselcvarela** [comment]
We could use `selectAmount` and use `-1` for `multiple`, `0` for `none` and then all positive numbers representing a limit of the selection. I think this is not necessary at the moment but can be useful on the future. For example, expecting only 3 items to be selected.

**Comment by rijkvanzanten** [comment]
@joselcvarela I don't think that necessarily solves for the "selectMode" prop 🤔

"selectMode" tells the layouts that we don't want to navigate to the detail page at all, and only want to select things. This is different from the amount of items that are allowed to be selected. We have to keep in mind that there's cases where you can both select X items _and_ navigate to the detail page, like on the tabular layout

**Comment by joselcvarela** [comment]
@nickrum 
@rijkvanzanten 

I was referring to `selectionMode` but I think what you are thinking.
So instead of having this 3 props, we could only have `selection` and `selectionMode`.
If `selectionMode` is:
- `none`. Clicking row opens the record and no checkboxes appear
- `one`. Clicking row selects it and radio buttons appear
- `multiple`. Clicking row selects it and checkboxes appear.

Seems correct?

**Comment by rijkvanzanten** [comment]
> Seems correct?

Not entirely! 😁 

When you open the collection page for any collection, the table layout allows both clicking the row to navigate, _and_ allows for selection multiple items:

CleanShot 2021-09-16 at 10 30 05@2x

(1) Will navigate to the detail page for the item
(2) Will select the item to enable batch edit/delete etc.

However, when the same layout is used in a relational selection drawer, the interaction change slightly...

CleanShot 2021-09-16 at 10 31 25@2x

Now, both (1) and (2) should select the row. The "one vs multiple" question is also only relevant in that drawer-use

**Comment by joselcvarela** [comment]
In my point of view, `selectMode` and `selectionMode` could be merged into a single prop called `selectMode` (like the first one).
This prop could take these as value:
- 'none': No checkboxes should appear. Clicking on rows should navigate
- 'readonly': Checkboxes should appear but are not modifiable.  Clicking on rows should do nothing
- 'one': Radio buttons should appear. Clicking on rows should select the current row and deselect previous selected row. Clicking on selecte row should deselect it.
- 'multiple': Checkboxes should appear. Clicking on rows should select/deselect according they are deselected/selected, respectively.

What do you think?

**Comment by rijkvanzanten** [comment]
> What do you think?

I think that still would have the same problem. For example, the tabular layout by default would have "multiple" _and_ needs to be able to click on a single row to navigate. By having it as a single prop, you can't have "multiple" and "none" from your examples at the same time

**Comment by joselcvarela** [comment]
@rijkvanzanten updated

**Comment by joselcvarela** [comment]
Here's a summary:

https://user-images.githubusercontent.com/14039341/134579587-68ea81a4-389f-45a1-866b-c9484a079138.mp4

File: app/src/layouts/tabular/tabular.vue [hunk 4]
```diff
@@ -7,7 +7,7 @@
 			v-model:headers="tableHeadersWritable"
 			class="table"
 			fixed-header
-			:show-select="readonly ? false : selection !== undefined"
+			:show-select="showSelect ? showSelect : selection !== undefined"
```
Comment (azrikahar) [inline app/src/layouts/tabular/tabular.vue:10]: Should we still include `readonly` in the logic here? Honestly not super sure when does "readonly" actually takes effect 🤔

**Comment by azrikahar** [comment]
## Update on additional changes I pushed

- some minor strict equality changes for consistency (`!=` to `!==`)
- Hid the checkbox for activity table, since it's what #7935 wanted 👍

## Question/Discussion

Show we hide the checkbox when the user only has `read` permission? The user won't be able to actually do anything with it.

eLFNanX8RZ

File: app/src/layouts/tabular/tabular.vue (hunk 4 above)
Comment (joselcvarela) [reply app/src/layouts/tabular/tabular.vue:10]: Hey @azrikahar 
Thank you for your review 🤝

It seems this `readonly` is only being used here:
https://github.com/directus/directus/blob/5d23dfd37f8b75039c0759820b0373a7ba459f97/app/src/modules/settings/routes/presets/item.vue#L10

I believe `readonly` could be used to prevent clicks on rows while `show-select` changes the visibility of checkbox.
IMO, it should be possible to have checkboxes but not be able to click on rows. The item in Presets is an example of this scenario.

**Comment by joselcvarela** [comment]
Thanks @azrikahar 
I think in the meantime the changes to fix #7935 have been lost.

That's a good question.
A use case of selection while having only read permissions is sharing links.
Since we do not have this feature for now, we can hide it, I think.

File: app/src/layouts/tabular/tabular.vue (hunk 4 above)
Comment (azrikahar) [reply app/src/layouts/tabular/tabular.vue:10]: ohh that makes perfect sense. Thanks for clarifying!

**Comment by azrikahar** [comment]
No worries 👍 The sharing link is interesting! Maybe leaving it on & show the checkboxes won't really be a huge deal since I assume a user _should_ understand they only have read permissions anyway. Maybe others can share more thoughts on it 😄

File: app/src/modules/notifications/routes/collection.vue [hunk 5]
```diff
@@ -8,6 +8,7 @@
 		:filter-user="filter"
 		:filter-system="roleFilter"
 		:search="search"
+		:show-select="'none'"
```
Comment (licitdev) [inline app/src/modules/notifications/routes/collection.vue:11]: Could remove the quotes for consistency.
```suggestion
		show-select="none"
```

**Comment by licitdev** [comment]
Pushed a fix to de-selecting by clicking on a row during selection mode.
The `item` was not reactive, thus the equality check in filtering does not work.

https://user-images.githubusercontent.com/26413686/141465030-eb5ab395-f088-41ae-8c7a-8e040bffb952.mov

There's a warning about the collection name not being a string. I believe it's something to do with the following code, but am unsure how to go about it.

https://github.com/joselcvarela/directus/blob/af2a8d4812345ec7080487d7ad93dcf7bf723cf0/app/src/modules/settings/routes/presets/item.vue#L332-L340

**Comment by Oreilles** [comment]
There are some issues with the map selection:
* Clicking on a point to select it doesn't work (it resets the selection but doesn't select the point)
* When holding the `alt` key while doing a selection, if some of the items you want to add are already selected, the whole selection is reset: (in the video the first 5 secs are holding `shift`, the following are holding `alt+shift`

https://user-images.githubusercontent.com/33065839/141978428-eac2aa76-198e-40c1-affb-5e1370c88f3d.mov

* I don't think we need to force additive behavior when multiple selection is allowed:  https://github.com/directus/directus/blob/fb5d3f8a75ac53c3137b730db25d7f5d6236dcb9/app/src/layouts/map/components/map.vue#L262

I suggest reverting to the previous behavior rules (but open to discuss it)
* Clicking on a point or doing a box-select should set the selection to only the clicked / hovered items, even if they already are selected
* Clicking on a point or doing a box-select  with `alt` pressed should add the clicked / hovered items to the selection if they aren't selected yet, and remove them from the selection it they already are.

Nothing to add on the rest of this PR, looks good!

**Comment by joselcvarela** [comment]
> * Clicking on a point or doing a box-select  with `alt` pressed should add the clicked / hovered items to the selection if they aren't selected yet, and remove them from the selection it they already are.

I think this should not be like this. I didn't know how to deselect a point and needed to check the source code to see how I could deselect it. When you see something selected/checked, usually you click on it for deselect.

What are your thoughts @benhaynes @rijkvanzanten ?

**Comment by rijkvanzanten** [comment]
Good point.. I'm actually not too sure what I would expect to happen 🤔 There's a bit of a conflict here between click to navigate and click to select. That being said, once you've selected one or more items, I would expect further clicks to toggle between the selection state (this would also match the UX from tabular/cards)

**Comment by benhaynes** [comment]
I'm a little lost in terms of what we're doing now, but here's what I think...

### Navigate

* When not in a selection state, clicking an item will navigate to the item detail page

### Select

* Select Mode: Clicking an item should toggle the individual selection state
* Select Mode: Dragging over items should **select** them, never deselect
* Select Mode: Clicking/dragging outside of items should deselect all of them
* Select Mode: Holding `alt` **could** temporarily enter **deselection** mode for clicking/dragging items
    * Ideally we also change the cursor so this is clear!
    * If we can't change the cursor (or something else) then this might not be clear enough
* Normal Mode: Holding `alt` should temporarily enter selection mode for clicking items
    * Ideally we also change the cursor so this is clear!

Basically, `alt` temporarily inverts the action. And we could add a note to the screen to tell the users about this keystroke.

Another point, when you drag OVER a point (but haven't released), it would be nice if it changed the style/state of the item dot to show it is about to be selected.

Did I miss anything? This is how most design applications work, and seems to be the most intuitive.

**Comment by Oreilles** [comment]
> Normal Mode: Holding alt should temporarily enter selection mode for clicking items

OK with these rules, @joselcvarela I can take it from there if that works for you!

**Comment by joselcvarela** [comment]
Yeah, no problem @Oreilles 🤝

**Review by rijkvanzanten — APPROVED** [review]

**Comment by rijkvanzanten** [comment]
It seems like the above mentioned rules by @benhaynes work in this PR (with Shift instead of Alt for mac vs windows interoperability). Seeing this PR fixes numerous bugs elsewhere, lets pick up any optimizations in the selection state of the map layout we'd wanna do in a separate PR.
